graphql2/graphqlapp/alert.go

```diff
@@ -177,6 +179,95 @@ func (q *Query) mergeFavorites(ctx context.Context, svcs []string) ([]string, er
 	return svcs, nil
 }
 
+// splitRangeByDuration splits the timeframe given between since and until by the duration provided.
+// Each segment is then transformed into an AlertDataPoint based on the given alerts.
+// The given alerts are required to be sorted by their CreatedAt field.
+func splitRangeByDuration(since, until time.Time, dur timeutil.ISODuration, alerts []alert.Alert) (result []graphql2.AlertDataPoint) {
+	// serialized as empty array
+	result = []graphql2.AlertDataPoint{}
+
+	if since.After(until) {
+		return result
+	}
+
+	// fill in timestamps
+	t1 := since
+	for t1.Before(until) {
+		result = append(result, graphql2.AlertDataPoint{
+			Timestamp: t1,
+		})
+		t1 = t1.AddDate(dur.Years, dur.Months, dur.Days).Add(dur.TimePart)
+	}
+
+	if len(result) == 0 {
+		return result
+	}
+
+	i := 0
+	// ignore alerts created before since
+	for i < len(alerts) && alerts[i].CreatedAt.Before(since) {
+		i++
+	}
+
+	for ptindex := range result {
+		upperBound := until
+		if ptindex != len(result)-1 {
+			upperBound = result[ptindex+1].Timestamp
+		}
+
+		for i < len(alerts) && alerts[i].CreatedAt.Before(upperBound) {
+			result[ptindex].AlertCount++
+			i++
+		}
+	}
```

Comment:
If we sort the alerts, can we just walk through, trimming the slice?

Something like:
```go
	countAlerts:=func(ts time.Time) int {
		var count int
		for i,a:=range alerts {
			if !a.CreatedAt.Before(ts) {
				alerts = alerts[i:]
				return count
			}
			if !a.CreatedAt.Before(until) {
				alerts=nil
				return count
			}

			count ++
		}

		alerts = nil
		return count
	}

	ts:=since
	// trim alerts before since
	countAlerts(ts)
	for ts.Before(until) {
		result=append(result, graphql2.AlertDataPoint{
			Timestamp: ts,
			AlertCount: countAlerts(ts),
		})
		ts=ts.AddDate(dur.Years, dur.Months, dur.Days).Add(dur.TimePart)
	}
```

Reply:
I think this looks good too. I especially like this part:

```
result=append(result, graphql2.AlertDataPoint{
			Timestamp: ts,
			AlertCount: countAlerts(ts),
		})
```

I think that's very clear. It returns an object with these fields and if a field needs to be calculated we pull that out like the countalerts function so this loop keeps it's clarity. Works for me +1.